### DISCUSSION TITLE
DDA and DIA identification using entrapment

### DISCUSSION THREAD

**Discussion by mlocardpaulet** [Proteobench/ProteoBench#373] [Resolved discussions leading to active modules]
Since we decided to give up the entrapment strategy, we discussed a different type of DDA identification module:
It would be nice to plot overlap of identification from pairs of searches. It is too complicated to have it as a main figure: we would need to select pairs. So we think of a main figure to select from
- Number of ID vs rank, we could have a slider that select the peptides matched to proteins identified with 1, 2, etc.. peptides; or 2 plots (proteins with >1 proteotypic peptide or all)
- Number of ID vs AUC (based on number of ID per FDR). Apparently we only need the q-value per psm in the output.
What do we count? Ions? Peptidoforms? PSMs would be easier, but what do we do if we have multiple PSMs per spectrum: We keep the best one. 
- Number of peptidoform sequence vs number or validated spectra
What file to use? Human sample?

**Comment by julianu** [comment]
Dataset for testing:
https://ftp.pride.ebi.ac.uk/pride/data/archive/2022/02/PXD028735/LFQ_Orbitrap_DDA_Human_01.raw

**Comment by julianu** [comment]
We should record the way of post-processing
- "simple" target decoy, percolator, other re-scoring

**Comment by veitveit** [comment]
Shouldn't be the level (PSM, peptidoform, protein) of the validation be on what was uploaded? 

For example, if you want to check for protein IDs, then the user should upload a table with protein accession names + their scores.

Otherwise, ProteoBench carries out part of the data analysis, and thus would influence the outcome by whatever choices have been made.

**Reply by julianu** [reply]
I would not really care about proteins for this module.
We definitely need the user to upload PSMs, I would suggest already with calculated q-values. Alternatively, we could also simply accept the PSMs of targets, decoys and the search engine score and calculate the scores ourself.

The benefit with already calculated q-values would be, we can also accept re-scored or percolated data.
But having a little "sanity check" would be good. Otherwise, one could just upload a totally synthetic, overtuned dataset.

**Comment by julianu** [comment]
A first impression of possible metrics (Dataset not run with the discussed FASTA, but using cRAP for contaminants)

![image](https://github.com/user-attachments/assets/aadc1edb-88ff-48c8-ba3f-a6ebd1058075)

![image](https://github.com/user-attachments/assets/92658e09-deb5-4361-88d0-1c185719523c)

![image](https://github.com/user-attachments/assets/6de3ee65-f58e-4c39-99ca-5d078bb41f16)

**Reply by julianu** [reply]
For now, the upset plot uses all datapoint. But these should probably be selected and allow only up to 5 datasets or something. Otherwise, this will get extremely messy.

**Reply by mlocardpaulet** [reply]
I would be more interested in the overlaps between all and all but one workflow. I personally would focus on these (if we have to choose).

**Reply by julianu** [reply]
Updates of the plots:

Plots with AUC from more "real" ROC (normalized by number of IDs per searchengine)
![image](https://github.com/user-attachments/assets/a6cb436c-a34c-4ddb-bc5c-8b3efc92061e)

And a plot of the number of IDs against q-value (not normalized)
![image](https://github.com/user-attachments/assets/a64e79e0-670d-400b-b217-7a69c7c66912)

... and normalized
![image](https://github.com/user-attachments/assets/aeb40319-8c80-4735-8fea-f0bb146dc491)

**Comment by mlocardpaulet** [comment]
I put here the QQ-plot paper that was suggested in another discussion: https://pubs.acs.org/doi/10.1021/acs.jproteome.2c00423  (discussion https://github.com/orgs/Proteobench/discussions/205 now closed to be discussed here)

**Comment by mlocardpaulet** [comment]
We had a very productive discussion at the Lorentz Workshop on Trustworthiness in Proteomics on a **ProteoBench module that would estimate random discovery proportion using entrapment**.
![IMG_2177](https://github.com/user-attachments/assets/18dcbc1b-8aef-4a3c-ba39-8fb054f20f50)
The idea is to use entrapment search a la Wen et al. https://www.nature.com/articles/s41592-025-02719-x.pdf

# General idea:
- The module will assess global FDR estimates provided by the tools for sequence identifications (or peptidoform? -> to explore based on what is possible with tools outputs).
Module design:
## Input data:
- Single data set with the choice between species (for example, for the DDA bulk data, we plan to have one raw file ecoli, one raw file yeast and one human, and to provide the corresponding fastas; users can upload the outputs file by file -- indicating the species of origin): 
   - bulk human and single cell.
   - Initially Thermo Astral (from https://www.biorxiv.org/content/10.64898/2026.01.29.702266v1.full.pdf).
   - in Raw (vendor) format files.
- Protein FASTA (human + contaminants) and entrapment FASTA is provided to participants.
## Metric
1. Use FDRBench (https://github.com/Noble-Lab/FDRBench) to create entrapment DB that all participants use.
  - Potential problem is that it creates a peptide database, not protein-level
  -> Reassemble? Not trivial maybe with missed cleavages. After quick testing, we seem to be able to derive the peptide pairing from the protein-level fasta output. Ask Bo Wen for help if this does not work. Use mimic (https://github.com/percolator/mimic) as fallback.
2. Reimplement entrapment FDR calculations from FDRBench (Java) in Python
## Participants must provide:
- Identification results in CSV/TSV/mzTab file thresholded at X% FDR (X≤1)
- and optionally: full list of unthresholded peptides (for more in-depth plots); Sequence/peptidoform q-values as reported by the tool
- Full parameter set
## To define:
- what parameters we want to parse and show in plots/tables
- Find a name for this module

**Comment by mlocardpaulet** [comment]
**TODO:** One thing that remains to be discussed/test:
do we count sequences/peptidoforms? 

1. figure out what tools provide output at peptide / peptidoform level?
2. Related to previous: should we keep only one row per peptidoform, keeping the best score?

**Comment by Cajac102** [comment]
Based on the discussion we had at the proteobench online meeting last week:

- Test data: 3 DIA single species (human) raw files from Barts Benchmarking paper
- Fasta file: Human proteome from UniProtKB + matched entrapment (1 per protein, scrambled, but I think the cleavage sites are kept)
- Focus on DIA precursor ion level for now -> we might need new parsers 
- Main plot: Showing #IDs against eFDR at 1% reported precursor level FDR
- In depth plot: Showing eFDR against reported precursor level FDR

 I will start setting up the module based on these decisions. If you object to any of them, let me know asap :)

**Reply by mlocardpaulet** [reply]
Perfect, thanks a lot! contaminants should also be in the Fasta ;)

**Reply by Cajac102** [reply]
I might have misunderstood the comment about which level this analysis should be on. Both DIANN and FragPipe only give precursor and protein level FDR, no peptide level. FDRBench also only does precursor and protein level, no in between. I'll go on with the precursor level for now, we can add a protein level module later.

**Reply by julianu** [reply]
I think if you are analyzing DIA data, then it is actually precursor level. Only on DDA we might need to differentiate between PSMs and peptidoforms (peptides with their modifications, and maybe also charges).

**Reply by Cajac102** [reply]
Update: The choice of which level to do this module on is a bit more tricky than I thought. FDRBench does do precursor, peptide, and protein level. However, there is also a difference between run-level FDR and global FDR, so if we use all 3 files, we also have to be careful which columns to use from which search engines. Based on an overview table I let ChatGPT do, the level that most DIA search engines report is the global protein level FDR.

<img width="1408" height="973" alt="image" src="https://github.com/user-attachments/assets/821766ef-3eff-463f-bfdd-85fb486dcf93" />

Imo it therefore makes the most sense to start with a module that checks if the global protein level FDR reported by these search engines falls within the entrapment calculated lower and upper boundaries.

**Reply by Cajac102** [reply]
Of course we can add other levels later, it would then be a matter of checking for each search engine if it calculates the FDR on that level and which column it reports it in.

**Reply by mlocardpaulet** [reply]
But then you would need new parsers for the protein-level tools outputs, right?

**Reply by Cajac102** [reply]
the search setting parsers are the most difficult, and these would stay the same. As for the output, we might need to adapt slightly in terms of which files to read, but that seems like a doable thing. For DIA-NN, we can still use the report.tsv for example.

**Reply by Cajac102** [reply]
For the main plot: there is no single value for the FDR based on FDRBench, they rather calculate upper and lower bounds. They propose 3 different ways (lower bound, upper bound ('combined'), another less conservative upper bound ('paired')). Based on where the FDR lands, the result is defined as 'invalid' (lower bound is higher than FDR), 'inconclusive' (FDR is between upper and lower bound), and 'valid' (upper bound is lower than FDR). 

Suggestion for the main plot:

<img width="3809" height="2751" alt="IMG_0028" src="https://github.com/user-attachments/assets/bc0e21a5-d1cc-49f9-9f3a-d71061f2907d" />

That would mean users do not have to interpret themselves if the calculated FDR is valid, it is immediately clear by the grouping. Might lead to cluttering within categories if two or more runs have a similar number of IDs, but we could add a bit of jitter.

**Reply by mlocardpaulet** [reply]
My understanding what that the "paired" eFDR was closest to the FDP. And the plan was to use it for the horizontal axis

**Reply by mlocardpaulet** [reply]
They even show that paired is super close to FDP On a sample of known proteins (https://www.nature.com/articles/s41592-025-02719-x.pdf). I like your plot idea, but it would not inform on how far off the inconclusive and invalid workflows are. Or how conservative the valid workflows are. 

We could also have some point-range type plot? Where the dot would be the paired eFDR, and there would be horizontal lines going from the lower bound to the upper bound values? But it may make the plot too complicated...

**Reply by Cajac102** [reply]
That was the first idea I started with, but indeed it will look complicated and hard to interpret:
<img width="4032" height="3024" alt="IMG_0029" src="https://github.com/user-attachments/assets/4f38086a-031b-47aa-93d9-58a8672b5a17" />

With the other plot, you could still check how a specific FDR estimation falls into the bounds in the FDR vs. eFDR plot in the in depth tab. 

If we only use the paired method in the main plot, users won't be able to distinguish between inconclusive and invalid FDR calculations.

**Reply by mlocardpaulet** [reply]
For this you could use colors or shapes. Use the paired eFDR in the horizontal axis, with 3 different colors if it is inconclusive, valid or else.

**Reply by Cajac102** [reply]
Hmmm maybe we can use the shapes for the categories and keep the colors for the search engines. Hard to make that plot colorblindfriendly then though.

**Reply by julianu** [reply]
I like the idea to give a "shift from paired/upper/lower" bound and the number of IDs. Interestingly, for the DDA Test of this data, the plots rather looked like this:
<img width="211" height="211" alt="image" src="https://github.com/user-attachments/assets/b52ffd9a-11e7-4b95-b218-6ddf23716b51" />
Showing the FDR against FDP for lower and upper bound as well as paired. Shown is MaxQuant and Oktoberfest, but this looked similar for all, just slightly shifted. So without taking care of the pairs, it would be inconclusive in the example. But with using the info, it actually is pretty valid at 1% (and basically all other values).

**Reply by Cajac102** [reply]
That's the decision we have to do in terms of interpreting the outcome of an entrapment experiment. I agree with you Julian that the FDR calculation shown looks quite solid - it is between a computed upper and lower bound. However in the paper, they see this scenario still as a potential 'strike' against the tool, as the 'actual FDP could be larger than the controlled FDR threshold (i.e., above the diagonal, invalid) or smaller than the controlled FDR threshold (i.e., below the diagonal, valid)' (see https://github.com/Noble-Lab/FDRBench/issues/14#issuecomment-4623595792). 
Maybe you discussed this during the meeting today though?

**Reply by julianu** [reply]
Actually, that's also what I saw in another dataset. Maybe just need to test a bit around - and also test our calculations on the dataset from the FDRBench manuscript.

**Comment by mlocardpaulet** [comment]
Update following a discussion in an online meeting (some new stuff, some that were already discussed here before):

- Quick concept of this ProteoBench module:
	1. We provide input data that everyone can download from the module documentation (here, it would be DIA raw files of a Human sample described [here](https://www.biorxiv.org/content/10.64898/2026.01.29.702266v2) and a fasta file that contains human sequences, contaminants, and a set of entrapped protein sequences generated with FDRBench). 
	2. Anyone can run a workflow for precursors/peptides identification
	3. Users can then upload to ProteoBench their analysis result and the module will calculate the empirical FDR. The module displays the results in what we call a “main figure”, where each workflow output (from a given tool, with a given set of parameters) is plotted. Additional tabs in the web application allow users to explore in more depths the results of each workflow.

- **What to plot in the main figure?** We concluded that the figure could be the flollowing: 
	1. Vertical axis = empirical FDR (calculated with the paired entrapped sequences generated with FDRBench) divided by the estimated FDR (FDR validation threshold of the workflow). (or shall we call them “FDP”?)
	2. Horizontal axis = workflows, maybe grouped by software tool or a given set of parameters? This will need a bit of testing. Caro had the nice idea to split them between inconclusive, valid and invalid, maybe we could try it? Or have a selection option to reorder based on parameters/criteria? Or can we find another metric that we would like to plot in this axis?
	3. Regarding the level of the analysis (**peptide or precursor?**): we’ll need to have both. Different tools perform validation at different levels. So we need to have one plot for precursor-level validation, and another for peptide level. Question: do we want to have them in the same module or do we need to develop 2 modules?

- Ideally, users would send us the entire set of spectrum-to-spectrum matches (SSMs), with the scores (no FDR-based filter). We all agree on that, but we believe that this would restrict too much the panel of tools that we could compare (since it is not always possible to get the full export). As a workaround, we would like to set up 2 options to upload workflow runs to the module: 
	1. The filtered set of peptides/precursors, so that we can calculate metrics in the main figure (with no other plots possible)
	2. The full list of SSMs with the scores. This would allow to calculate more metrics and provide more figures. 

- We shortly discussed keeping protein-N-terminal methionine in the entrapped sequences because a lot of software tools include N-ter methionine excision as default during the search.

- **Shall we consider PTMs?** Tricky. Bill suggested that we have a look at this [paper](https://pubs.acs.org/doi/10.1021/acs.jproteome.3c00399), and that we discuss this with Uri. I think that the best idea we came up with for now was: require that users perform the search without any variable PTM. We all agreed that this was not ideal, but did not come up with a better workable plan…

**Reply by mlocardpaulet** [reply]
Follow up on this discussion: 

Idea for a summary metric proposed by Ville: 

We could consider using the area under the curve? For example, in the image from the [FDRBench github page](https://github.com/Noble-Lab/FDRBench):
<img width="2700" height="2700" alt="image" src="https://github.com/user-attachments/assets/e6f9e015-5177-4ba0-9358-ba003b855aa1" />

The green line initially hugs the diagonal, then deviates slightly. The area between the green line and the diagonal is > 0, and it can be calculated or estimated using standard numerical integration methods. If the green line were entirely below the diagonal, probably you want to set the area under the curve to zero. Then zero, or close to zero, indicates a valid method.
If you only have, say, 10 data points, then an equivalent metric could be the sum of the excess. For example, if estimated FDP (green) is 6.3% at 6% FDR threshold, then the excess at that point is 6.3% - 6% = 0.003. Sum the excess over the 10 points and you get a number. If it's >0, the method is inconclusive or not valid. Makes sense?

This would only be possible for the "full" submission where users provide all PSMs/SSMs with scores.

**Comment by Cajac102** [comment]
-- MOCK DATA! --

Here are some mockups for potential main plots:

1) Vanilla: # IDs vs. [lower bound OR paired OR combined]

<img width="1132" height="645" alt="Screenshot 2026-06-16 at 10 42 07" src="https://github.com/user-attachments/assets/aa248e9d-53f5-4c50-aa8f-8421653dd9eb" />

Negative:
- Confusing: Should upper bound be lower or higher - I would not know how to interpret this plot 
- only works with 1% (or any other fixed threshold) across all points, as there is no comparison to the reported FDR
- does not spark any joy in me

Positive:
+ shows # IDs 
+ scalable with lots of points
+ similar to our usual main plots


2) Pretty Forests: FDP Interval Plot

<img width="1133" height="503" alt="Screenshot 2026-06-16 at 14 54 39" src="https://github.com/user-attachments/assets/3f43514e-a37d-4a7e-972f-a652a01bdfa3" />


Negative:
- Does not scale well with lots of points (will get longer and longer)
- Does not show # IDs (could be implemented by e.g. scaling the size of the marker. edit: now in clear text next to label)
- Hard to make colorblind friendly as color is used to categorise valid invalid inconclusive

Positive:
+ Clearly shows lower bound, upper bound, and reported, therefore easy to interpret
+ sparks joy

3) Get ratio'd: FDP / FDR

<img width="1169" height="504" alt="Screenshot 2026-06-16 at 10 42 16" src="https://github.com/user-attachments/assets/26b4b714-8c5a-4074-8a7b-d15d54d36272" />

Negative:
- Confusing: why is one point orange and one is red, when they are both above the line (due to only one bound shown)
- Turning FDR or FDP into ratio makes it a bit more abstract
- Does not show # IDs (could be implemented, e.g. by scaling point size)

Positive:
- can also use other thresholds than 1% 
- Scalable with lots of points

4) Get ratio'd x Vanilla: 

<img width="1159" height="540" alt="Screenshot 2026-06-16 at 11 45 42" src="https://github.com/user-attachments/assets/19f228dc-243f-404f-9c69-c36d54c99897" />

Negative:
- Confusing: why is one point invalid and one is inconclusive, when they are both above the line (due to only one bound shown)
- Turning FDR or FDP into ratio makes it a bit more abstract
- Telling apart colors and shapes takes a couple of mental CPU cycles

Positive:
- can also use other thresholds than 1% 
- Scalable with lots of points
- Shows # IDs

5) Guinea Flag: 

<img width="1157" height="511" alt="Screenshot 2026-06-16 at 11 56 40" src="https://github.com/user-attachments/assets/3f745a68-be37-4553-ab7d-0c3afd9f793c" />

Negative
- unclear how conservative the FDR calculation is, as only the three categories are shown

Positive:
- very interpretable

**Reply by Cajac102** [reply]
Current fan favorite:
Show two main plots:
1) Get ratio'd x Vanilla 
2) Pretty Forests underneath

We could maybe under certain circumstances potentially... connect the two. You click on one scatterpoint, you get the bar highlighted in the forest plot.

**Comment by mlocardpaulet** [comment]
We had a few meetings with interested contributors to refine the design of the module. Here are the meetings notes:

- Quick concept of this ProteoBench module:
1. We provide input data that everyone can download from the module documentation (here, it would be DIA raw files of a Human sample described [here](https://url.au.m.mimecastprotect.com/s/NmdfCNLJyQUEq6MgGu4hJcy5_DK?domain=urldefense.com) and a fasta file that contains human sequences, contaminants, and a set of entrapped protein sequences generated with FDRBench). 
2. Anyone can run a workflow for precursors/peptides identification
3. Users can then upload to ProteoBench their analysis result and the module will calculate the empirical FDR. The module displays the results in what we call a “main figure”, where each workflow output (from a given tool, with a given set of parameters) is plotted. Additional tabs in the web application allow users to explore in more depths the results of each workflow.
An alpha version of the module with some mock data can be found here: https://proteobench.cubimed.rub.de/Entrapment_DIA_ion_Astral.

- For the moment, the module calculates metrics at precursor ions level. The plan is to have another one at peptide level, and we will stop here for now, waiting to have a better strategy for protein group FDR benchmarking. For both modules (precursor and peptide), we use peptide-level shuffling for the database generation.

- We had a discussion on the FDR threshold at which the input data should be filtered. Matthew would like to have all the precursors (so results of a search with 100% FDR) and filter at a given FDR percentage ourselves. We think that this makes sense, but software could have some other post-processing magic that lead to our filtering at 1% being different from the software output filtered at 1%. So the conclusion was to keep it flexible: users can provide results filtered at the threshold they want. We can totally encourage them to provide all precursors. 

- Order of the main figures? We think that to cover all aspects of FDR evaluation, we need to provide several figures. Our two favorite (that we discussed more) were the “forest plot” that is the first one on the current page, and is accompanied with a horizontal barplot that shows the number of identified features; and the scatter plot below showing the number of identified features (vertical axis) and the upper or lower bound (horizontal axis: its default value should be upper bound, it can be changed with a selection field). The point shape can correspond to valid, invalid, inconclusive. Please have a look and tell us which one you prefer. We’ll keep both but need to decide which one to put first. I think that we all agreed that the forest plot conveys the information we think important very nicely, but risk to become overcrowded. For this reason, and for clarity, we should always have a button that allows to plot the metrics at a given FDR.  

- In the figures, we propose to have a filter to set the FDR threshold that we want to apply to the results. This means that for all the search outputs with a FDR threshold above the one that is selected for the plot, we will report the estimated FDP for the same FDR target as all other workflows. 

- How do we handle modifications? For this point, I copy/paste here Uri’s comments from his email:
	- Regarding Robbin's comment about the fact that disabling PTMs might increase false discoveries because a PTM-generated spectrum might match to a non-modified wrong peptide (I think Matthew was making an argument along the same lines). I agree that the risk is there, and it’d certainly diminish the tool’s power. However, I think it’s also valid to demand that a tool controls the FDR when searching without PTMs even if there are PTMs in the sample, so I don’t think this can invalidate the entrapment protocol.
	- Robbe raised the point that it might create an issue when people might forget to switch off PTMs and that’s also a good point, but I think that we can address this by offering two modes: with PTMs and w/o, each with its own page and set of figures, and by simply looking at the list of reported peptides we can determine whether PTMs were enabled or not, and route the results to the right page.
	- When PTMs are disabled things are fairly straightforward and similar to peptide level analysis.
	- When PTMs are enabled, the lower bound is still valid, albeit it’s even more conservative than it normally is. The upper bound needs an asterisk which we can display as a clear message, preferably somewhere on the top of the page, along the lines of “Note that the current entrapment construction does not account for localization errors. For example, the tool might report one localization instead of another, or might report two separate localizations whereas the sample only contains one. None of these errors is currently properly accounted for."
	- What do you think?

- One more question that we did not have time to discuss. If you have a strong opinion I am interested: how shall we call the entrapment-based FDR estimation metric that we calculate? 
   - FDP
   - eFDR
  - RDP

- We shortly discussed keeping protein-N-terminal methionine in the entrapped sequences because a lot of software tools include N-ter methionine excision as default during the search.

**Comment by mlocardpaulet** [comment]
Follow up on our weekly ProteoBench meeting: 
- we can call the metric “estimated FDP”
- for modifications, we will go with the strategy proposed by Uri: parameters will be set up by users, and if they add variable modifications to the search we will add a note/warning stating that localisation errors are not properly accounted for.

**Comment by mlocardpaulet** [comment]
Hi @Cajac102, really great work!
In the "main" scatter plot, I wonder if we should keep the choice to visualise the "Maximum reported" Q-value threshold. I think that it can be too different between points, and it is not super easy to find in the plot what these parameters are set to. Maybe this could be missleading?

<img width="887" height="482" alt="image" src="https://github.com/user-attachments/assets/13fe8d62-1a92-4278-b94d-6d7f7bda4866" />

**Comment by Cajac102** [comment]
we could remove it, the reason it exists is for cases where people upload FDRs that are not in the list and want to see the point anyways at that exact FDR. But that will probably never happen anyways. Ill remove it.
